File: R/repeat.R
#' Get or set the repeat count for an `openmpt` module
#' 
#' Tracker music can be composed such that it is intended to play
#' in a continuous loop. With the repeat count you can affect the number of
#' times a module is repeated when playing with [`play()`] or rendered with [`convert_mod()`].
#' @param mod A tracker module object of class `openmpt`.
#' @param ... Ignored
#' @param value An `integer` value to assign to the repeat count.
#' @returns Returns the `integer` repeat count of an `openmpt` object. In case
#' of an assign operator (`<-`) an updated `mod` is returned with the new repeat count.
#' @examples
#' mod <- demo_mod()
#' repeat_count(mod) <- 2
#' repeat_count(mod)
#' @rdname repeat
#' @export
repeat_count <- function(mod, ...) {
  get_repeat_count_(mod)
}

#' @rdname repeat
#' @export
`repeat_count<-` <- function(mod, ..., value) {
  set_repeat_count_(mod, as.integer(value))
}

File: src/repeat.cpp
#include <cpp11.hpp>
#include <libopenmpt/libopenmpt.hpp>
using namespace cpp11;
using namespace openmpt;

module * get_mod(SEXP mod); // specified in get_mod.cpp

[[cpp11::register]]
int get_repeat_count_(SEXP mod) {
  module * my_mod = get_mod(mod);
  return my_mod->get_repeat_count();
}

[[cpp11::register]]
SEXP set_repeat_count_(SEXP mod, int repeat_count) {
  module * my_mod = get_mod(mod);
  try {
    my_mod->set_repeat_count(repeat_count);
  } catch(...) {
    cpp11::stop("Failed to set repeat_count");
  }
  return mod;
}
